core/src/main/java/org/elasticsearch/cluster/routing/RoutingChangesObserver.java

```diff
@@ -0,0 +1,196 @@
+
+    /**
+     * Called when a shard is failed or cancelled.
+     */
+    void shardFailed(ShardRouting failedShard, UnassignedInfo unassignedInfo);
+
+    /**
+     * Called on relocation source when relocation completes after relocation target is started.
+     */
+    void relocationCompleted(ShardRouting removedRelocationSource);
+
+    /**
+     * Called on replica relocation target when replica relocation source fails. Promotes the replica relocation target to ordinary
+     * initializing shard.
+     */
+    void relocationSourceRemoved(ShardRouting removedReplicaRelocationSource);
+
+    /**
+     * Called on started primary shard after it has been promoted from replica to primary and is reinitialized due to shadow replicas.
+     */
+    void startedPrimaryReinitialized(ShardRouting startedPrimaryShard, ShardRouting initializedShard);
+
+    /**
+     * Called when started replica is promoted to primary.
+     */
+    void replicaPromoted(ShardRouting replicaShard);
+
+
+    /**
+     * Abstract implementation of {@link RoutingChangesObserver} that does not take any action. Useful for subclasses that only override
+     * certain methods.
+     */
+    class AbstractRoutingChangesObserver implements RoutingChangesObserver {
+
+        @Override
+        public void shardInitialized(ShardRouting unassignedShard) {
+
+        }
+
+        @Override
+        public void shardStarted(ShardRouting initializingShard, ShardRouting startedShard) {
+
+        }
+
+        @Override
+        public void relocationStarted(ShardRouting startedShard, ShardRouting targetRelocatingShard) {
+
+        }
+
+        @Override
+        public void unassignedInfoUpdated(ShardRouting unassignedShard, UnassignedInfo newUnassignedInfo) {
+
+        }
+
+        @Override
+        public void shardFailed(ShardRouting activeShard, UnassignedInfo unassignedInfo) {
+
+        }
+
+        @Override
+        public void relocationCompleted(ShardRouting removedRelocationSource) {
+
+        }
+
+        @Override
+        public void relocationSourceRemoved(ShardRouting removedReplicaRelocationSource) {
+
+        }
+
+        @Override
+        public void startedPrimaryReinitialized(ShardRouting startedPrimaryShard, ShardRouting initializedShard) {
+
+        }
+
+        @Override
+        public void replicaPromoted(ShardRouting replicaShard) {
+
+        }
+    }
+
+    class DelegatingRoutingChangesObserver implements RoutingChangesObserver {
+
+        private final RoutingChangesObserver[] routingChangesObservers;
+
+        public DelegatingRoutingChangesObserver(RoutingChangesObserver... routingChangesObservers) {
+            this.routingChangesObservers = routingChangesObservers;
+        }
+
+        @Override
+        public void shardInitialized(ShardRouting unassignedShard) {
+            for (RoutingChangesObserver routingChangesObserver : routingChangesObservers) {
+                routingChangesObserver.shardInitialized(unassignedShard);
```

Comment:
This is executed on the cluster state update thread and used to compute the next cluster state. The code executed here is supposed to be side-effect free and the cluster state update operation should always be all-or-nothing: No partial failures are tolerated, otherwise this might lead in a inconsistent cluster state. The snapshot/restore observer will just update the RestoreInProgress entries in the cluster state to bring them to a consistent state with the shard routing entries in the routing table, similar as what is done here with the MetaDataUpdater.